project5/3700kvstore

```diff
@@ -514,96 +582,140 @@ class Server(LoggingMixin):
+        """
         if self.state == State.LEADER:
-            # Respond with the value if it exists
-            self.send_ok(msg, value=self.datastore.get(msg['key'], ''))
-        elif self.leader == 'FFFF':
-            self.failed.add('{}.get'.format(msg['src']))
-            self.send_fail(msg)
+            if msg['type'] == 'put':    
+                for e in self.committed_entries:
+                    if e['MID'] == msg['MID']:
+                        self.send_ok(msg, value=self.get(e['key']))
+                        break
+                else:
+                    # Append entry to local log
+                    key = msg['key']
+                    value = msg.get('value', '')
+                    entry = {
+                        'command': msg['type'],
+                        'src': msg['src'],
+                        'MID': msg['MID'],
+                        'key': key,
+                        'value': value,
+                        'term': self.current_term
+                    }
+                    self.log.append(entry)
+                    self.info('{} {}.{} ACCEPTED'.format(msg['MID'], msg['src'], msg['type']))
+            else:
+                self.idx_get_list.append( GetEntry(self.commit_idx, msg, current_time()) )
+                # If we have no outstanding entries, send out a no-op to give the gets
+                # something to trigger off of
+                if self.commit_idx == self.last_log_idx:
+                    self.append_empty()
+        # Send fail if leader unknown
         else:
-            self.info('{}.get redirected to {}'.format(msg['src'], self.leader))
-            self.send_redirect(msg)
+            self.fail_or_redirect[msg['MID']] = msg
 
     def apply(self, entries):
         """Commits the entries to the state machine"""
         for e in entries:
             if e['command'] == 'put':
-                self.datastore[e['key']] = e['value']
+                self.put(e['key'], e['value'])
 
     # ------------------------ END OF CLIENT INTERFACE ------------------------
 
-    def run(self):
-        """Runs the main loop"""
-        while True:
-            time = current_time()
-            # Leaders send empty AppendEntries RPCs (heartbeats) to each 
-            # server during idle periods to prevent election timeouts
-            if self.state == State.LEADER:
-                span = time - self.prev_heartbeat
-                if span >= HEARTBEAT_FREQ:
-                    self.debug('Sent Heartbeat')
-                    self.broadcast_heartbeat()
-                    # Reset the heartbeat timer
-                    self.prev_heartbeat = time
+    def manage_fail_or_redirect(self, time):
+        fail_or_redirect_span = time - self.fail_or_redirect_timer
+        if self.state == State.LEADER:
+            for msg in self.fail_or_redirect.values():
+                self.dispatch(msg['type'], msg) 
+            self.fail_or_redirect = {}
```

Comment:
You can pull out the `self.fail_or_redirect = {}` to the end of the function since it occurs in every case.

Reply:
That last one in an elseif not an else, the default case here performs no action